bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/RowResultAdapter.java

```diff
@@ -0,0 +1,224 @@
+import com.google.protobuf.ByteString;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.client.Result;
+
+/**
+ * Adapter for {@link RowAdapter} that uses {@link Result} to represent logical rows.
+ *
+ * <p>For internal use only - public for technical reasons.
+ */
+@InternalApi("For internal usage only")
+public class RowResultAdapter implements RowAdapter<Result> {
+
+  private static final byte[] EMPTY_VALUE = new byte[0];
+
+  @Override
+  public RowBuilder<Result> createRowBuilder() {
+    return new RowResultBuilder();
+  }
+
+  @Override
+  public boolean isScanMarkerRow(Result result) {
+    if (result instanceof RowResult) {
+      return ((RowResult) result).isMarkerRow();
+    }
+    // This may never be executed still is it ok to leave it here.
+    return result.isEmpty();
+  }
+
+  @Override
+  public ByteString getKey(Result result) {
+    if (result instanceof RowResult) {
+      return ((RowResult) result).getKey();
+    }
+    return ByteStringer.wrap(result.getRow());
+  }
+
+  static class RowResult extends Result {
+    private final ByteString rowKey;
+    private final boolean isMarkerRow;
+
+    RowResult(ByteString rowKey, List<Cell> cells) {
+      super();
+      this.rowKey = rowKey;
+      this.isMarkerRow = cells == null || cells.isEmpty();
+
+      // all except default ctor of Result are private, So instantiating cells through copyFrom()
+      // because value(), size(), isEmpty() rawCells() etc. are directly using Result's cells field.
+      this.copyFrom(Result.create(cells));
+    }
+
+    ByteString getKey() {
+      return rowKey;
+    }
+
+    boolean isMarkerRow() {
+      return isMarkerRow;
+    }
+  }
+
+  public static class RowResultBuilder implements RowBuilder<Result> {
+    private ByteString currentKey;
+    private String family;
+    private byte[] qualifier;
+    private List<String> labels;
+    private long timestamp;
+    private byte[] value;
+
+    private Map<String, List<RowCell>> cells = new TreeMap<>();
+    private List<RowCell> currentFamilyCells = null;
+    private String previousFamily;
+    private int nextValueIndex;
+
+    @Override
+    public void startRow(ByteString rowKey) {
+      this.currentKey = rowKey;
+    }
+
+    @Override
+    public void startCell(
+        String family, ByteString qualifier, long timestamp, List<String> labels, long size) {
+      this.family = family;
+      this.qualifier = qualifier.toByteArray();
+      this.timestamp = timestamp;
+      this.labels = labels;
+      if (size > 0) {
+        this.value = new byte[(int) size];
+        this.nextValueIndex = 0;
+      } else {
+        this.value = EMPTY_VALUE;
+        this.nextValueIndex = -1;
+      }
+    }
+
+    @Override
+    public void cellValue(ByteString newValue) {
```

Comment:
The current implementation doesn't really make sense to me.

Can I suggest something like this instead?

```java
    public void cellValue(ByteString newValue) {
      // Optimize unsplit cells by avoiding a copy
      if (nextValueIndex == -1) {
        this.value = ByteStringer.extract(newValue);
        return;
      }

      Preconditions.checkState(nextValueIndex + newValue.size() <= value.length,
              "Cell value is larger than expected");

      newValue.copyTo(this.value, this.nextValueIndex);
      nextValueIndex += newValue.size();
    }
```